src/sql-parser/src/parser.rs

```diff
@@ -1648,13 +1643,64 @@ impl<'a> Parser<'a> {
-            Some(self.parse_format()?)
+        // legacy upsert format syntax allows setting the key format after the keyword UPSERT, so we
+        // may mutate this variable in the next block
+        let mut format = if matches!(connector, Connector::Kafka { .. }) {
+            // pure.rs asserts that KeyValue formats only make sense for Kafka
+            match self.parse_one_of_keywords(&[KEY, FORMAT]) {
+                Some(KEY) => {
+                    self.expect_keyword(FORMAT)?;
+                    let key = self.parse_format()?;
+                    self.expect_keywords(&[VALUE, FORMAT])?;
+                    let value = self.parse_format()?;
+                    CreateSourceFormat::KeyValue { key, value }
+                }
+                Some(FORMAT) => CreateSourceFormat::Bare(self.parse_format()?),
+                Some(_) => unreachable!("parse_one_of_keywords returns None for this"),
+                None => CreateSourceFormat::None,
+            }
+        } else if self.parse_keyword(FORMAT) {
+            CreateSourceFormat::Bare(self.parse_format()?)
         } else {
-            None
+            CreateSourceFormat::None
         };
         let envelope = if self.parse_keyword(ENVELOPE) {
-            self.parse_envelope()?
+            let envelope = self.parse_envelope()?;
+            if matches!(envelope, Envelope::Upsert) {
+                // TODO: remove support for explicit UPSERT FORMAT after a period of deprecation
+                if self.parse_keyword(FORMAT) {
+                    warn!("UPSERT FORMAT has been deprecated, use the new KEY FORMAT syntax");
+                    if let CreateSourceFormat::Bare(value) = format {
+                        let key = self.parse_format()?;
+                        format = CreateSourceFormat::KeyValue { key, value };
+                    } else {
+                        self.error(
+                            self.index,
+                            "Key/Value format specified multiple times".into(),
+                        );
+                    }
+                } else if format.is_simple() {
+                    // the existing semantics of Upsert is that specifying a simple bare format
+                    // duplicates the format into the key.
+                    //
+                    // TODO(bwm): We should either make this the semantics everywhere, or deprecate
+                    // this.
+                    let value = format.value();
+                    if let Some(value) = value {
+                        format = CreateSourceFormat::KeyValue {
+                            key: value.clone(),
+                            value,
+                        }
+                    } else {
+                        return parser_err!(
+                            self,
+                            self.peek_pos(),
+                            "Upsert requires a value format"
+                        );
+                    };
+                }
```

Comment:
I don't think it makes sense to do this normalization in the parser. (I.e., this last `format.is_simple()` branch.) Normalizing the `ENVELOPE UPSERT` into the new `KEY FORMAT` seems ok because it will facilitate the migration, but this normalization looks like it will stick around a while and it really seems like semantic analysis to me. Like, the planner should just do this interpretation of `CreateSourceFormat::Bare`.